runtime/src/main/java/org/corfudb/runtime/collections/ScopedTransaction.java

```diff
@@ -0,0 +1,211 @@
+package org.corfudb.runtime.collections;
+
+import com.google.protobuf.Message;
+import lombok.extern.slf4j.Slf4j;
+import org.corfudb.protocols.wireprotocol.Token;
+import org.corfudb.runtime.CorfuRuntime;
+import org.corfudb.runtime.CorfuStoreMetadata.Timestamp;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.BiFunction;
+import java.util.function.BiPredicate;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+@Slf4j
+public class ScopedTransaction implements
+        StoreTransaction<TxnContext>, CommitApi, AutoCloseable {
+
+    private final TransactionCrud<TxnContext> crud;
+    private final CorfuRuntime runtime;
+    private final Token txnSnapshot;
+    private final Map<
+            Table<? extends Message, ? extends Message, ? extends Message>,
+            Table> mapping;
+
+    ScopedTransaction(
+            @Nonnull final CorfuRuntime runtime,
+            @Nonnull final String namespace,
+            @Nonnull final IsolationLevel isolationLevel,
+            Table<?, ?, ?>... tables) {
+        this.crud = new TransactionCrud<>(namespace, this::txAbort);
+        this.runtime = runtime;
+        this.txnSnapshot = txBeginInternal(isolationLevel);
+        this.mapping = Arrays.stream(tables).collect(Collectors.toMap(
+                        table -> table,
+                        table -> table.generateImmutableView(txnSnapshot.getSequence())));
+    }
+
+    private Token txBeginInternal(IsolationLevel isolationLevel) {
+        log.trace("TxnContext: begin transaction in namespace {}", crud.namespace);
+
+        if (isolationLevel.getTimestamp() != Token.UNINITIALIZED) {
+            return isolationLevel.getTimestamp();
+        }
+
+        return runtime
+                .getSequencerView()
+                .query()
+                .getToken();
+    }
+
+    @Override
+    public Timestamp commit() {
+        throw new UnsupportedOperationException("This is a read only transaction.");
```

Comment:
Rather than have a method which throws `UnsupportedOperationException` why not just avoid having this method all together?

Reply:
Sounds good.